### PR TITLE
Add support for custom JS embeds in the App

### PR DESCRIPTION
**Pull request by br41nslug**
## Description

This PR adds the `embed` function for `hook` extensions. These embeds allow for integration with 3rd party scripts. 

> Warning: as this feature allows for direct HTML injection it is very possible to break the App using this improperly!

> Warning: You will still need to allow any external domains in the CSP for these embeds to work properly in the browser!

## Examples
GTM integration example (https://developers.google.com/tag-platform/tag-manager/web) with `GTM_ID` as environment variable.
```js
export default function registerHook({ embed }, { env }) {
  embed('head', () => `<!-- Google Tag Manager -->
    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);})(window,document,'script','dataLayer','${env.GTM_ID}');</script>
    <!-- End Google Tag Manager -->`);
};
```
Sentry integration example (https://docs.sentry.io/platforms/javascript/install/cdn/)
```js
export default function registerHook({ embed }, { env }) {
  embed('head', '<script src="https://browser.sentry-cdn.com/7.21.1/bundle.min.js" integrity="sha384-xOL2QebDu7YNMtC6jW2i5RpQ5RcWOyQMTwrWBiEDezpjjXM7mXhYGz3vze77V91Q" crossorigin="anonymous"></script>')
  embed('body', () => `<script>
  Sentry.init({
    dsn: "${env.SENTRY_DSN}" // "https://examplePublicKey@o0.ingest.sentry.io/0",
    release: "my-project-name@${env.npm_package_version}",
    integrations: [new Sentry.BrowserTracing()],
  
    // We recommend adjusting this value in production, or using tracesSampler
    // for finer control
    tracesSampleRate: 1.0,
  });
  </script>`);
};
```

Fixes ENG-19

## Type of Change

- [ ] Bugfix
- [ ] Improvement
- [X] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [X] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [x] Documentation was added/updated. PR: https://github.com/directus/docs/pull/235

### CODE REVIEW

**Review by wrynegade — COMMENTED** [review]
Seems like a clever approach to embeds! Always have to be careful with snippet injection, but including an extension and updating CSP both require elevated privileges so seems okay in the current context.

If we want to load extensions from sources other than the file system, we'll want to make sure things like embeds are limited to trusted sources or only allowed from the file system.

**Comment by piotr-cz** [comment]
Isn't env part of define hook interface?

File: api/src/extensions.ts [hunk 1]
```diff
@@ -488,6 +498,19 @@ class ExtensionManager {
...
+				const content = typeof code === 'function' ? code(env) : code;
+				if (content.trim().length === 0) {
+					logger.warn(`Couldn't register embed hook. Provided code is empty!`);
+					return;
+				}
+				if (position === 'head') {
+					this.hookEmbedsHead.push(content);
+				}
+				if (position === 'body') {
+					this.hookEmbedsHead.push(content);
```
Comment (piotr-cz) [inline api/src/extensions.ts:511]: Shouldn't this be `this.hookEmbedsBody.push(content)` ?

Comment (br41nslug) [reply api/src/extensions.ts:511]: It definitely should, thanks 😄

File: api/src/extensions.ts [hunk 2]
```diff
@@ -500,7 +499,7 @@ class ExtensionManager {
 				}
 			},
 			embed: (position: 'head' | 'body', code: string | EmbedHandler) => {
-				const content = typeof code === 'function' ? code(env) : code;
+				const content = typeof code === 'function' ? code() : code;
```
Comment (piotr-cz) [inline api/src/extensions.ts:502]: What's the point of using function now?
Sorry for being PITA

Comment (br41nslug) [reply api/src/extensions.ts:502]: Nothing wrong with asking critical questions 😄 it has lost most of it's utility when not being used to passed the env however I do like that this allows you to create a function scope for this hook (similar to the others) so you don't have to do any variable declaration or conditional logic in the root extension scope (or separate named function).
I think for both consistency in hooks and developer experience it still adds something.

Comment (piotr-cz) [reply api/src/extensions.ts:502]: Fair enough, this makes sense.
Thank you

**Comment by paescuj** [comment]
There's a great library called [Partytown](https://github.com/BuilderIO/partytown) which could be interesting in this context:
> Partytown is a lazy-loaded library to help relocate resource intensive scripts into a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API), and off of the [main thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread). Its goal is to help speed up sites by dedicating the main thread to your code, and offloading third-party scripts to a web worker.

Haven't looked at the details yet though, but maybe it would be worth considering an integration or we could just mention it in the documentation along with this feature.

**Review by licitdev — APPROVED** [review]
Work well with a clean implementation! 👍 

However, I'm wondering if a comment should be added to surround the custom embed blocks for easy identification, which allows for monitoring.

**Comment by paescuj** [comment]
> However, I'm wondering if a comment should be added to surround the custom embed blocks for easy identification, which allows for monitoring.

I like this idea! 👍

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**